File: sdk-js/src/agent.ts
import { Conduit, defaultClient } from "./client.js";
import { Policy } from "./policy.js";
import type {
  AgentJSON,
  Balance,
  CreateAgentOptions,
  InvoiceJSON,
  PayOptions,
  ReceiptJSON,
  TransactionJSON,
} from "./types.js";

/** Generate a fresh idempotency key. Prefers Web Crypto (Node 20+ and browsers). */
export function newIdempotencyKey(): string {
  const c = (globalThis as { crypto?: Crypto }).crypto;
  if (c && typeof c.randomUUID === "function") return c.randomUUID();
  // Fallback for runtimes without Web Crypto (should not happen on Node >= 20).
  return `idem-${Date.now().toString(16)}-${Math.floor(Math.random() * 1e16).toString(16)}`;
}

export interface SendOptions {
  idempotencyKey?: string;
}

export interface Receipt {
  id: string;
  agentId: string;
  status: "pending" | "settled" | "failed";
  hash: string | null;
  amountSats: number;
  feeSats: number;
  platformFeeSats: number;
  settledInMs: number | null;
  destination: string | null;
  memo: string | null;
  createdAt: Date;
}

export interface Invoice {
  id: string;
  agentId: string;
  paymentRequest: string;
  paymentHash: string;
  amountSats: number;
  memo: string | null;
  status: "pending" | "settled" | "failed";
  expiresAt: Date;
  createdAt: Date;
}

export interface Transaction {
  id: string;
  agentId: string;
  direction: "send" | "receive";
  amountSats: number;
  feeSats: number;
  platformFeeSats: number;
  destination: string | null;
  paymentHash: string | null;
  status: "pending" | "settled" | "failed";
  memo: string | null;
  settledAt: Date | null;
  latencyMs: number | null;
  createdAt: Date;
}

export interface LedgerAdjustment {
  agentId: string;
  transactionId: string;
  /** positive = credit, negative = debit */
  deltaSats: number;
  balanceSats: number;
}

export interface LedgerAdjustJSON {
  agent_id: string;
  transaction_id: string;
  delta_sats: number;
  balance_sats: number;
}

export function fromLedger(d: LedgerAdjustJSON): LedgerAdjustment {
  return {
    agentId: d.agent_id,
    transactionId: d.transaction_id,
    deltaSats: d.delta_sats,
    balanceSats: d.balance_sats,
  };
}

export function fromReceipt(r: ReceiptJSON): Receipt {
  return {
    id: r.id,
    agentId: r.agent_id,
    status: r.status,
    hash: r.hash,
    amountSats: r.amount_sats,
    feeSats: r.fee_sats,
    platformFeeSats: r.platform_fee_sats ?? 0,
    settledInMs: r.settled_in_ms,
    destination: r.destination,
    memo: r.memo,
    createdAt: new Date(r.created_at),
  };
}
export function fromInvoice(i: InvoiceJSON): Invoice {
  return {
    id: i.id,
    agentId: i.agent_id,
    paymentRequest: i.payment_request,
    paymentHash: i.payment_hash,
    amountSats: i.amount_sats,
    memo: i.memo,
    status: i.status,
    expiresAt: new Date(i.expires_at),
    createdAt: new Date(i.created_at),
  };
}
export function fromTx(t: TransactionJSON): Transaction {
  return {
    id: t.id,
    agentId: t.agent_id,
    direction: t.direction,
    amountSats: t.amount_sats,
    feeSats: t.fee_sats,
    platformFeeSats: t.platform_fee_sats ?? 0,
    destination: t.destination,
    paymentHash: t.payment_hash,
    status: t.status,
    memo: t.memo,
    settledAt: t.settled_at ? new Date(t.settled_at) : null,
    latencyMs: t.latency_ms,
    createdAt: new Date(t.created_at),
  };
}

export class Agent {
  readonly id: string;
  readonly name: string;
  readonly pubkey: string | null;
  active: boolean;
  readonly createdAt: Date;
  readonly policy: Policy;
  private readonly client: Conduit;

  private constructor(data: AgentJSON, client: Conduit) {
    this.id = data.id;
    this.name = data.name;
    this.pubkey = data.pubkey;
    this.active = data.active;
    this.createdAt = new Date(data.created_at);
    this.client = client;
    this.policy = new Policy(this, client);
  }

  static async create(opts: CreateAgentOptions, client?: Conduit): Promise<Agent> {
    const c = client ?? defaultClient();
    const data = await c.post<AgentJSON>("/v1/agents", {
      name: opts.name,
      daily_limit: opts.dailyLimit,
      metadata: opts.metadata,
    });
    return new Agent(data, c);
  }

  static async get(id: string, client?: Conduit): Promise<Agent> {
    const c = client ?? defaultClient();
    const data = await c.get<AgentJSON>(`/v1/agents/${id}`);
    return new Agent(data, c);
  }

  static async list(client?: Conduit): Promise<Agent[]> {
    const c = client ?? defaultClient();
    const data = await c.get<{ data: AgentJSON[] }>("/v1/agents");
    return data.data.map((d) => new Agent(d, c));
  }

  async pay(opts: PayOptions): Promise<Receipt> {
    const data = await this.client.post<ReceiptJSON>(
      "/v1/payments/pay",
      {
        agent_id: this.id,
        to: opts.to,
        sats: opts.sats,
        memo: opts.memo,
        metadata: opts.metadata,
      },
      { idempotencyKey: opts.idempotencyKey ?? newIdempotencyKey() },
    );
    return fromReceipt(data);
  }

  async sendInvoice(
    paymentRequest: string,
    opts: { sats?: number; memo?: string; idempotencyKey?: string } = {},
  ): Promise<Receipt> {
    const data = await this.client.post<ReceiptJSON>(
      "/v1/payments/send",
      {
        agent_id: this.id,
        payment_request: paymentRequest,
        sats: opts.sats,
        memo: opts.memo,
      },
      { idempotencyKey: opts.idempotencyKey ?? newIdempotencyKey() },
    );
    return fromReceipt(data);
  }

  async keysend(
    destPubkey: string,
    sats: number,
    memo?: string,
    opts: SendOptions = {},
  ): Promise<Receipt> {
    const data = await this.client.post<ReceiptJSON>(
      "/v1/payments/send",
      {
        agent_id: this.id,
        dest_pubkey: destPubkey,
        sats,
        memo,
      },
      { idempotencyKey: opts.idempotencyKey ?? newIdempotencyKey() },
    );
    return fromReceipt(data);
  }

  async receive(amount: number, opts: { memo?: string; expiry?: number } = {}): Promise<Invoice> {
    const data = await this.client.post<InvoiceJSON>("/v1/invoices", {
      agent_id: this.id,
      amount,
      memo: opts.memo,
      expiry: opts.expiry ?? 3600,
    });
    return fromInvoice(data);
  }

  /** Operator top-up: credit this agent's virtual balance (admin scope). */
  async credit(
    sats: number,
    opts: { reason?: string; metadata?: Record<string, unknown> } = {},
  ): Promise<LedgerAdjustment> {
    const body: Record<string, unknown> = { sats };
    if (opts.reason !== undefined) body.reason = opts.reason;
    if (opts.metadata !== undefined) body.metadata = opts.metadata;
    return fromLedger(
      await this.client.post<LedgerAdjustJSON>(`/v1/agents/${this.id}/credit`, body),
    );
  }

  /** Operator sweep: debit this agent's virtual balance (admin scope). */
  async debit(
    sats: number,
    opts: { reason?: string; metadata?: Record<string, unknown> } = {},
  ): Promise<LedgerAdjustment> {
    const body: Record<string, unknown> = { sats };
    if (opts.reason !== undefined) body.reason = opts.reason;
    if (opts.metadata !== undefined) body.metadata = opts.metadata;
    return fromLedger(
      await this.client.post<LedgerAdjustJSON>(`/v1/agents/${this.id}/debit`, body),
    );
  }

  async balance(): Promise<Balance> {
    const data = await this.client.get<{
      available_sats: number;
      pending_sats: number;
      total_sats: number;
    }>(`/v1/agents/${this.id}/balance`);
    return {
      available: data.available_sats,
      pending: data.pending_sats,
      total: data.total_sats,
    };
  }

  async transactions(limit = 50, direction?: "send" | "receive"): Promise<Transaction[]> {
    const data = await this.client.get<{ data: TransactionJSON[] }>(
      `/v1/agents/${this.id}/transactions`,
      direction ? { limit, direction } : { limit },
    );
    return data.data.map(fromTx);
  }

  async deactivate(): Promise<void> {
    await this.client.delete(`/v1/agents/${this.id}`);
    this.active = false;
  }
}

File: sdk-python/conduit/agent.py
from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime
from typing import Any

from .client import Conduit, default_client
from .invoice import Invoice
from .payment import Receipt, _parse_dt
from .policy import Policy
from .transaction import Transaction


def _new_idempotency_key() -> str:
    return str(uuid.uuid4())


@dataclass
class Balance:
    available: int
    pending: int
    total: int


class Agent:
    """An autonomous wallet with optional spending policy.

    Mirrors the website code panel:

        agent = Agent.create(name="compute-router-7", daily_limit=50_000)
        agent.policy.attach(max_per_hour=10_000, allowlist=["02beef..."])
        receipt = agent.pay(to="compute-node-7@lnd.conduit.energy",
                            sats=150, memo="dataset query")
    """

    def __init__(
        self,
        *,
        id: str,
        name: str,
        pubkey: str | None,
        active: bool,
        created_at: datetime,
        client: Conduit | None = None,
    ) -> None:
        self.id = id
        self.name = name
        self.pubkey = pubkey
        self.active = active
        self.created_at = created_at
        self._client = client or default_client()
        self.policy = Policy(self, client=self._client)

    # --- factory methods ---

    @classmethod
    def create(
        cls,
        name: str,
        daily_limit: int | None = None,
        metadata: dict[str, Any] | None = None,
        *,
        client: Conduit | None = None,
    ) -> "Agent":
        c = client or default_client()
        payload: dict[str, Any] = {"name": name}
        if daily_limit is not None:
            payload["daily_limit"] = daily_limit
        if metadata is not None:
            payload["metadata"] = metadata
        data = c.post("/v1/agents", json=payload)
        return cls._from_api(data, client=c)

    @classmethod
    def get(cls, agent_id: str, *, client: Conduit | None = None) -> "Agent":
        c = client or default_client()
        data = c.get(f"/v1/agents/{agent_id}")
        return cls._from_api(data, client=c)

    @classmethod
    def list(cls, *, client: Conduit | None = None) -> list["Agent"]:
        c = client or default_client()
        data = c.get("/v1/agents")
        return [cls._from_api(item, client=c) for item in data.get("data", [])]

    @classmethod
    def _from_api(cls, data: dict[str, Any], client: Conduit) -> "Agent":
        return cls(
            id=data["id"],
            name=data["name"],
            pubkey=data.get("pubkey"),
            active=bool(data.get("active", True)),
            created_at=_parse_dt(data["created_at"]),
            client=client,
        )

    # --- actions ---

    def pay(
        self,
        *,
        to: str,
        sats: int,
        memo: str | None = None,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> Receipt:
        """Pay a Lightning address (`name@host`) or a BOLT11 invoice.

        An `Idempotency-Key` is sent automatically (a fresh UUID4 if you don't
        provide one) so an automatic retry can never double-pay. Pass an
        explicit `idempotency_key` to make a manual retry idempotent too.
        """
        data = self._client.post(
            "/v1/payments/pay",
            json={
                "agent_id": self.id,
                "to": to,
                "sats": sats,
                "memo": memo,
                "metadata": metadata,
            },
            idempotency_key=idempotency_key or _new_idempotency_key(),
        )
        return Receipt.from_api(data)

    def send_invoice(
        self,
        payment_request: str,
        *,
        sats: int | None = None,
        memo: str | None = None,
        idempotency_key: str | None = None,
    ) -> Receipt:
        """Pay a BOLT11 invoice (or zero-amount invoice with explicit sats)."""
        data = self._client.post(
            "/v1/payments/send",
            json={
                "agent_id": self.id,
                "payment_request": payment_request,
                "sats": sats,
                "memo": memo,
            },
            idempotency_key=idempotency_key or _new_idempotency_key(),
        )
        return Receipt.from_api(data)

    def keysend(
        self,
        dest_pubkey: str,
        sats: int,
        *,
        memo: str | None = None,
        idempotency_key: str | None = None,
    ) -> Receipt:
        data = self._client.post(
            "/v1/payments/send",
            json={
                "agent_id": self.id,
                "dest_pubkey": dest_pubkey,
                "sats": sats,
                "memo": memo,
            },
            idempotency_key=idempotency_key or _new_idempotency_key(),
        )
        return Receipt.from_api(data)

    def receive(
        self, amount: int, *, memo: str | None = None, expiry: int = 3600
    ) -> Invoice:
        """Create a Lightning invoice that other agents/services can pay."""
        data = self._client.post(
            "/v1/invoices",
            json={"agent_id": self.id, "amount": amount, "memo": memo, "expiry": expiry},
        )
        return Invoice.from_api(data)

    @property
    def balance(self) -> Balance:
        data = self._client.get(f"/v1/agents/{self.id}/balance")
        return Balance(
            available=int(data["available_sats"]),
            pending=int(data["pending_sats"]),
            total=int(data["total_sats"]),
        )

    def transactions(self, *, limit: int = 50, direction: str | None = None) -> list[Transaction]:
        params: dict[str, Any] = {"limit": limit}
        if direction:
            params["direction"] = direction
        data = self._client.get(f"/v1/agents/{self.id}/transactions", params=params)
        return [Transaction.from_api(item) for item in data.get("data", [])]

    def deactivate(self) -> None:
        self._client.delete(f"/v1/agents/{self.id}")
        self.active = False

    def __repr__(self) -> str:
        return f"<Agent id={self.id!r} name={self.name!r} active={self.active}>"
